keps/sig-storage/20200120-skip-permission-change.md

```diff
@@ -0,0 +1,147 @@
+creation-date: 2020-01-20
+last-updated: 2020-01-20
+status: implementable
+see-also:
+replaces:
+superseded-by:
+
+---
+
+# Skip Volume Ownership Change
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Graduation Criteria](#graduation-criteria)
+  - [Test Plan](#test-plan)
+  - [Monitoring](#monitoring)
+- [Implementation History](#implementation-history)
+- [Drawbacks [optional]](#drawbacks-optional)
+- [Alternatives [optional]](#alternatives-optional)
+- [Infrastructure Needed [optional]](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Summary
+
+Currently before a volume is bind-mounted inside a container the permissions on
+that volume are changed recursively to the provided fsGroup value.  This change
+in ownership can take an excessively long time to complete, especially for very
+large volumes (>=1TB) as well as a few other reasons detailed in [Motivation].
+To solve this issue we will add a new field called `pod.Spec.SecurityContext.VolumePermissionChangePolicy` and
+allow the user to specify whether they want the ownership change to occur.
+
+## Motivation
+
+When a volume is mounted on the node, we recursively change permissions of volume
+before bind mounting the volume inside container. The reason of doing this is to ensure
+that volumes are readable/writable by provided fsGroup.
+
+But this presents following problems:
+ - An application(many popular databases) which is sensitive to permission bits changing
+   underneath may refuse to start whenever volume being used inside pod gets mounted on
+   different node.
+ - If volume has a large number of files, performing recursive `chown` and `chmod`
+   could be slow and could cause timeout while starting the pod.
+
+### Goals
+
+ - Allow volume ownership and permission to be skipped during mount
+
+### Non-Goals
+
+ - In some cases if user brings in a large enough volume from outside, the first time ownership and permission change still could take lot of time.
+ - On SELinux enabled distributions we will still do recursive chcon whenever applicable and handling that is outside the scope.
+
+## Proposal
+
+We propose that an user can optionally opt-in to skip recursive ownership(and permission) change on the volume if volume already has right permissions.
+
+### Implementation Details/Notes/Constraints [optional]
+
+When creating a pod, we propose that `pod.Spec.SecurityContext` field expanded to include a new field called `VolumePermissionChangePolicy` which can have following possible values:
+
+ - `NoChange` --> Don't change permissions and ownership. This could be useful to users when security policy of a cluster prevents users from removing `fsGroup` from their pod's `SecurityContext`.
```

Comment:
When would you want this if fsGroup is set? Could this conflict with PodSecurityPolicy?

Reply:
I was thinking database use case I linked above. Some applications do not want permission bits (`rwx`) to change on files owned by them. It is fine for `fsGroup` to be set initially because otherwise application won't be able to write to the volume itself unless application is running as root, but once up and running with a fixed fsGroup the application does not want permission bits to change.

Reply:
So this isn't exactly NoChange. We still change the ownership, just not the permission bits?

What I'm wondering is what does it mean if an admin wants to require a certain fsgroup for this Pod, but then the Pod can actually override that?

Reply:
When I was thinking about `NoChange` - I did not intend this policy to cause any ownership or permission change. But I think you are right - we could use a policy that causes ownership change but not permission bits change.  This would satisfy use case I had in mind. what do you think?

Reply:
I have removed `NoChange` for now. The reason being - I do not think it is necessary even for use case I mentioned because applications mostly care about permissions on files/directories created by them, rather than top level directory and hence using `OnDemand` will work.